cpp/dolfinx/mesh/Topology.h

```diff
@@ -145,13 +205,25 @@ class Topology
   // MPI communicator
   dolfinx::MPI::Comm _comm;
 
-  // Cell type
-  CellType _cell_type;
-
-  // Parallel layout of entities for each dimension
-  std::array<std::shared_ptr<const common::IndexMap>, 4> _index_map;
-
-  // AdjacencyList for pairs [d0][d1] == d0 -> d1 connectivity
+  // Cell types for entites in Topology, as follows:
+  // [CellType::point, edge_types..., facet_types..., cell_types...]
+  // Only one type is expected for vertices, (and usually edges), but facets
+  // and cells can be a list of multiple types, e.g. [quadrilateral, triangle]
+  // for facets.
+  // Offsets are position in the list for each entity dimension, in
+  // AdjacencyList style.
+  std::vector<CellType> _entity_types;
+  std::vector<std::int8_t> _entity_type_offsets;
+
+  // Parallel layout of entities for each dimension and cell type
+  // flattened in the same layout as _entity_types above.
+  std::vector<std::shared_ptr<const common::IndexMap>> _index_map;
+
+  // Connectivity between entity dimensions and cell types, arranged as
+  // a 2D array. Indexing is 0=vertex, 1=edge, followed by [facet types], [cell
+  // types]. e.g. for a 3D mesh with tets, pyramid and hex: 0=vertex, 1=edge,
+  // 2=triangle, 3=quad, 4=tet, 5=pyr, 6=hex. Connectivity between different
```

Comment:
The text here should probably not be hard-coded, and rather refer to the CellType enum that we use for sorting, as this might change.

Reply:
clarified documentation here